### PR TITLE
CART-688 uri: URI changes/PSR changes stage1

### PR DESCRIPTION
**Pull request by frostedcmos**
URI lookup logic modified. Current mode:

- For clients, when attempting to lookup uri for rank:tag=x,
rank:tag=0 is first checked in the cache. If rank:tag=0 is present, it is used as
a target of URI lookup request for rank:tag=x. Otherwise psr_rank:tag=0
is used as a target for URI lookup request.
NOTE: This logic will further be modified/expanded in full CART-688 implementation.

- Servers are assumed to always know rank:tag=0 URIs for any rank
as tag=0 URI is what servers primary ranks are added with.
As such, servers will request rank:tag=x from rank:tag=0

- URI lookup handler on the server side will either return URI
for rank:tag=x or if not found (psr might not know all tag URIs of all servers)
it will return rank:tag=0 URI back. rank:tag=0 URI is guaranteed to be known
if the rank is part of the group; otherwise -DER_OOG is returned to the caller.

- ul_tag field of the output of the URI_LOOKUP request will indicate which tag the
uri was returned for.

As such for clients, when clients attempt to lookup URI of rank:tag=x
for the PSR:

- If a lookup request is made for rank:tag=x and instead rank:tag=0 is
returned by the server (as indicated by ul_tag output field), then
rank:tag=0 URI will be populated in the local cache and new URI request
going to rank:tag=0 will be issued to lookup rank:tag=x

Signed-off-by: Alexander Oganezov <alexander.a.oganezov@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
If you merge this with master now checkpatch should run correctly.

File: src/cart/src/cart/crt_group.c [hunk 1]
```diff
@@ -1638,45 +1539,33 @@ crt_hdlr_uri_lookup(crt_rpc_t *rpc_req)
...
 	/**
-	 * step 2, make sure URI of tag 0 is in cache in preparation to forward
-	 * the lookup RPC
+	 * step 2, if rank:tag is not found, lookup rank:tag=0
 	 */
-	if (ul_in->ul_tag != 0) {
-		/* tag != 0, check if tag 0 is in cache */
-		rc = crt_grp_lc_lookup(grp_priv_primary, crt_ctx->cc_idx,
+	ul_out->ul_tag = 0;
+	rc = crt_grp_lc_lookup(grp_priv_primary, crt_ctx->cc_idx,
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1555]: Alignment?

File: src/cart/src/cart/crt_group.c [hunk 2]
```diff
@@ -1638,45 +1539,33 @@ crt_hdlr_uri_lookup(crt_rpc_t *rpc_req)
...
-				grp_priv_primary->gp_pub.cg_grpid,
-				g_rank, 0, rc);
-			D_GOTO(out, rc);
-		}
+	if (rc != 0) {
+		D_ERROR("crt_grp_lc_lookup(grp %s, rank %d, tag %d) "
+			"failed, rc: %d.\n",
+			grp_priv_primary->gp_pub.cg_grpid,
+			g_rank, 0, rc);
+		D_GOTO(out, rc = -DER_OOG);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1562]: Always DER_OOG or something else?  What if rc was -DER_NO_MEM

File: src/cart/src/cart/crt_rpc.c [hunk 3]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
-
-	rpc_priv->crp_state = RPC_STATE_INITED;
-	rc = crt_req_send_internal(rpc_priv);
-
-out:
-	return rc;
-}
+static int
+crt_issue_uri_lookup(crt_context_t ctx, crt_group_t *group,
+			d_rank_t contact_rank, uint32_t contact_tag,
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/cart/src/cart/crt_rpc.c:541]: If these are current can you accept them, although checkpatch does have an issue where it only checks continued indentation on the second line, so L542 and L543 probably need changing as well.

File: src/cart/src/cart/crt_rpc.c [hunk 4]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
-		RPC_ERROR(rpc_priv,
-			  "failed cci_rc: %d\n",
-			  cb_info->cci_rc);
-		if (cb_info->cci_rc == -DER_OOG)
-			D_GOTO(out, rc = -DER_OOG);
-
-		/* TODO: Proper selection of new PSR is required */
-		rc = cb_info->cci_rc;
-		D_GOTO(out, rc);
+		D_ERROR("URI_LOOKUP rpc completed with rc=%d\n",
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:560]: Would RPC_ERROR work here?

File: src/cart/src/cart/crt_rpc.c [hunk 5]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
-	/* extract uri */
-	ul_out = crt_reply_get(cb_info->cci_rpc);
-	D_ASSERT(ul_out != NULL);
+	ul_in = crt_req_get(lookup_rpc);
+	ul_out = crt_reply_get(lookup_rpc);
 
-	if (ul_out->ul_rc != 0)
-		D_GOTO(out, rc = ul_out->ul_rc);
+	if (ul_out->ul_rc != 0) {
+		D_ERROR("URI_LOOKUP returned rc=%d\n", ul_out->ul_rc);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:569]: Again, RPC_ERROR.

File: src/cart/src/cart/crt_rpc.c [hunk 6]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
-		D_ERROR("crt_req_send_internal() failed, rc %d, opc: %#x\n",
-			rc, rpc_priv->crp_pub.cr_opc);
+		RPC_ERROR(chained_rpc_priv,
+			"crt_req_fill_tgt_uri() failed; rc=%d\n", rc);
 		D_GOTO(out, rc);
 	}
 
+	rc = crt_req_send_internal(chained_rpc_priv);
+retry:
+	/* TODO: add retry logic for CART-688 */
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:609]: Possibly log something here if there's missing functionality.

File: src/cart/src/cart/crt_rpc.c [hunk 7]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
+			D_GOTO(out, rc = -DER_NONEXIST);
+		}
 
-	rc = crt_req_create(rpc_priv->crp_pub.cr_ctx, &ul_tgt_ep,
-			    CRT_OPC_URI_LOOKUP, &ul_req);
-	if (rc != 0) {
-		D_ERROR("crt_req_create URI_LOOKUP failed, rc: %d opc: %#x.\n",
-			rc, rpc_priv->crp_pub.cr_opc);
+		rc = crt_issue_uri_lookup(ctx, grp,
+					lookup_rank, 0,
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/cart/src/cart/crt_rpc.c:684]: These look like they'd fit on the same line.

File: src/cart/src/cart/crt_group.c (hunk 2 above)
Comment (frostedcmos) [reply src/cart/src/cart/crt_group.c:1562]: Lookup actually no longer performs any allocations and doesn't/can't fail with rc. I ll change lookup to be void in next patch along with other changes.

File: src/cart/src/cart/crt_group.c (hunk 1 above)
Comment (frostedcmos) [reply src/cart/src/cart/crt_group.c:1555]: hm not sure what alignment you are referring to. it appears properly aligned within scope of block to me

File: src/cart/src/cart/crt_rpc.c (hunk 5 above)
Comment (frostedcmos) [reply src/cart/src/cart/crt_rpc.c:569]: will change

File: src/cart/src/cart/crt_rpc.c (hunk 4 above)
Comment (frostedcmos) [reply src/cart/src/cart/crt_rpc.c:560]: will change

File: src/cart/src/cart/crt_group.c (hunk 2 above)
Comment (ashleypittman) [reply src/cart/src/cart/crt_group.c:1562]: I like that kind of change ;)  Error-path handing is notoriously hard to test and there's a lot of error-paths that exist and we have code for like this one that can never actually be invoked anyway.

File: src/cart/src/cart/crt_group.c (hunk 1 above)
Comment (ashleypittman) [reply src/cart/src/cart/crt_group.c:1555]: The line below, but github wouldn't allow me to comment on that as it was unmodified.

Comment (frostedcmos) [reply src/cart/src/cart/crt_group.c:1555]: ah ok, will change

File: src/cart/src/cart/crt_rpc.c (hunk 6 above)
Comment (frostedcmos) [reply src/cart/src/cart/crt_rpc.c:609]: will add

File: src/cart/src/cart/crt_rpc.c [hunk 8]
```diff
@@ -536,251 +536,259 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
-		D_GOTO(out, rc);
+		D_ERROR("URI_LOOKUP rpc completed with rc=%d\n",
+			cb_info->cci_rc);
+		D_GOTO(retry, rc = cb_info->cci_rc);
 	}
 
-	/* extract uri */
-	ul_out = crt_reply_get(cb_info->cci_rpc);
-	D_ASSERT(ul_out != NULL);
+	ul_in = crt_req_get(lookup_rpc);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:565]: we generally prefer lookup_rpc->cr_input although in practice it doesn't make much difference.

**Pull request merged**